koordinator-sh/koordinator
# koord-scheduler: DeviceShare supports preempting devices

### Ⅰ. Describe what this PR does

Enhanced DeviceShare scheduling plugins
- support preempting Devices



### Ⅱ. Does this pull request fix one issue?



### Ⅲ. Describe how to verify it

1. Create a Pod and apply for all GPUs on a specified node. In the scenario I tested, a node has two GPU instances
```bash
$ cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test-reserve-gpu
  name: test-gpu-deploy
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: test-reserve-gpu
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: test-reserve-gpu
        koordinator.sh/qosClass: LS
    spec:
      containers:
      - args:
        - "3600"
        command:
        - sleep
        image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda10.2
        name: test
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
            koordinator.sh/gpu: "200"
      schedulerName: koord-scheduler
EOF
deployment.apps/test-gpu-deploy created
```
2. Check Pod status
```bash
$ kubectl get pod -o wide
NAME                               READY   STATUS    RESTARTS   AGE   IP         NODE                    NOMINATED NODE   READINESS GATES
test-gpu-deploy-849874876f-cfvrr   1/1     Running   0          10s   10.0.3.7   cn-beijing.10.0.3.245   <none>           <none>
``` 
3. Create a Pod with high priority than pre-created pod
```bash
$ cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: high-test-reserve-gpu
  name: high-test-gpu-deploy
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: high-test-reserve-gpu
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: high-test-reserve-gpu
        koordinator.sh/qosClass: LS
    spec:
      containers:
      - args:
        - "3600"
        command:
        - sleep
        image: nvcr.io/nvidia/k8s/cuda-sample:vectoradd-cuda10.2
        name: test
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
            koordinator.sh/gpu: "100"
      priorityClassName: system-cluster-critical
      schedulerName: koord-scheduler
EOF
deployment.apps/high-test-gpu-deploy created
```
5. Watch the Pod status
```bash
$ kubectl get pod -o wide
NAME                                   READY   STATUS        RESTARTS   AGE   IP         NODE                    NOMINATED NODE          READINESS GATES
high-test-gpu-deploy-dcd465c44-mq9bv   0/1     Pending       0          4s    <none>     <none>                  cn-beijing.10.0.3.245   <none>
test-gpu-deploy-849874876f-8fpsk       0/1     Pending       0          4s    <none>     <none>                  <none>                  <none>
test-gpu-deploy-849874876f-cfvrr       1/1     Terminating   0          21s   10.0.3.4   cn-beijing.10.0.3.245   <none>                  <none>
```
7.  Get events of preempted pod
```bash
$ kubectl get event | grep cfvrr
38s         Normal    Preempted           pod/test-gpu-deploy-849874876f-cfvrr        Preempted by default/high-test-gpu-deploy-dcd465c44-mq9bv on node cn-beijing.10.0.3.245
```
As the result shows, Pod `test-gpu-deploy-849874876f-cfvrr` is preempted by `high-test-gpu-deploy-dcd465c44-mq9bv`.

```bash
$ kubectl get pod 
NAME                                   READY   STATUS    RESTARTS   AGE
high-test-gpu-deploy-dcd465c44-mq9bv   1/1     Running   0          6m39s
test-gpu-deploy-849874876f-8fpsk       0/1     Pending   0          6m38s
```

### Ⅳ. Special notes for reviews

### V. Checklist

- [x] I have written necessary docs and comments
- [x] All checks passed in `make test`
